### PR TITLE
Read/Take return NO_DATA if a future change is found [12134]

### PR DESCRIPTION
**Pull request by richiware**
`ReadTakeCommand` breaks the search of new changes if it found a future change of a writer although there is unread changes from other writers.

### CODE REVIEW

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
How about a regression test?

**Comment by richiware** [comment]
@MiguelCompany regression test added

File: test/unittest/dds/subscriber/CMakeLists.txt [hunk 1]
```diff
@@ -43,6 +43,8 @@ if(NOT ((MSVC OR MSVC_IDE) AND EPROSIMA_INSTALLER))
 
         add_executable(DataReaderTests ${DATAREADERTESTS_SOURCE})
         target_compile_definitions(DataReaderTests PRIVATE FASTRTPS_NO_LIB
+            BOOST_ASIO_STANDALONE
+            ASIO_STANDALONE
```
Comment (MiguelCompany) [inline test/unittest/dds/subscriber/CMakeLists.txt:47]: No need of these, as we are linking fastrtps (see my next comment)

File: test/unittest/dds/subscriber/DataReaderTests.cpp [hunk 2]
```diff
@@ -57,6 +57,9 @@
 
 #include "../../logging/mock/MockConsumer.h"
 
+#include <rtps/transport/test_UDPv4Transport.h>
```
Comment (MiguelCompany) [inline test/unittest/dds/subscriber/DataReaderTests.cpp:60]: This is adding an unnecessary dependency on asio (and thus the build fails), but only the transport descriptor is necessary:

```suggestion
#include <fastdds/rtps/transport/test_UDPv4TransportDescriptor.h>
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: test/unittest/dds/subscriber/DataReaderTests.cpp (hunk 2 above)
Comment (richiware) [reply test/unittest/dds/subscriber/DataReaderTests.cpp:60]: I need access to `rtps::test_UDPv4Transport::test_UDPv4Transport_ShutdownAllNetwor` in line 1974.

File: test/unittest/dds/subscriber/CMakeLists.txt (hunk 1 above)
Comment (richiware) [reply test/unittest/dds/subscriber/CMakeLists.txt:47]: Without that compiler is trying to find a boost header file.

File: test/unittest/dds/subscriber/DataReaderTests.cpp (hunk 2 above)
Comment (MiguelCompany) [reply test/unittest/dds/subscriber/DataReaderTests.cpp:60]: You could achieve the same with `percentageOfMessagesToDrop` or `messages_filter_` on `test_UDPv4TransportDescriptor`

Comment (richiware) [reply test/unittest/dds/subscriber/DataReaderTests.cpp]: Done

File: test/unittest/dds/subscriber/CMakeLists.txt (hunk 1 above)
Comment (richiware) [reply test/unittest/dds/subscriber/CMakeLists.txt]: Done

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Pull request merged**